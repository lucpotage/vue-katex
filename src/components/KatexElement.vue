<template>
  <component :is="tag" v-html="math" />
</template>

<script>
export default {
  name: 'KatexElement',
}
</script>

<script setup>
// Would like to use v-dompurify-html in the component instead of v-html in the template.
// But I haven't because it changes the test answers.
import { computed, inject, toRefs, ref } from 'vue'
import katex from 'katex'
import merge from 'deepmerge'

const removeUndefined = (obj) => {
  const newObj = {}
  for (const key of Object.keys(obj)) {
    if (typeof obj[key] !== 'undefined') {
      newObj[key] = obj[key]
    }
  }
  return newObj
}

const props = defineProps({
  expression: {
    type: String,
    default: '',
    required: true,
  },
  displayMode: {
    type: Boolean,
    default: undefined,
  },
  throwOnError: {
    type: Boolean,
    default: undefined,
  },
  errorColor: {
    type: String,
    default: undefined,
  },
  macros: {
    type: Object,
    default: undefined,
  },
  colorIsTextColor: {
    type: Boolean,
    default: undefined,
  },
  maxSize: {
    type: Number,
    default: undefined,
  },
  maxExpand: {
    type: Number,
    default: undefined,
  },
  allowedProtocols: {
    type: Array,
    default: undefined,
  },
  strict: {
    type: [Boolean, String, Function],
    default: undefined,
  },
})

const {
  displayMode,
  expression,
  throwOnError,
  errorColor,
  macros,
  colorIsTextColor,
  maxSize,
  maxExpand,
  allowedProtocols,
  strict,
} = toRefs(props)

const globalOptions = ref({})
globalOptions.value = inject('$katexOptions')

const options = computed(() => {
  return merge(
    globalOptions.value,
    removeUndefined({
      displayMode: displayMode.value,
      throwOnError: throwOnError.value,
      errorColor: errorColor.value,
      macros: macros.value,
      colorIsTextColor: colorIsTextColor.value,
      maxSize: maxSize.value,
      maxExpand: maxExpand.value,
      allowedProtocols: allowedProtocols.value,
      strict: strict.value,
    })
  )
})

const tag = computed(() => {
  return displayMode.value ? 'div' : 'span'
})
const math = computed(() => {
  return katex.renderToString(expression.value, options.value)
})
</script>
