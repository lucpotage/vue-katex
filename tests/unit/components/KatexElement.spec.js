import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import KatexElement from '@/components/KatexElement.vue'
import VueKatex from '@/plugin.js'
import katex from 'katex'

describe('KatexElement.vue', () => {
  it('matches snapshot - inline mode', () => {
    const wrapper = mount(KatexElement, {
      props: {
        expression: '\\frac{a_i}{1+x}',
      },
      global: {
        plugins: [VueKatex],
      },
      shallow: true,
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('matches snapshot - display mode', () => {
    const wrapper = mount(KatexElement, {
      props: {
        expression: '\\frac{a_i}{1+x}',
        displayMode: true,
      },
      global: {
        plugins: [VueKatex],
      },
      shallow: true,
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('respects global options', () => {
    const wrapper = mount(KatexElement, {
      props: {
        expression: '\\frac{a_i}{1+x}',
      },
      global: {
        plugins: [
          [
            VueKatex,
            {
              globalOptions: {
                macros: {
                  '\\blah': '\\frac{#}{#}',
                },
              },
            },
          ],
        ],
        shallow: true,
      },
    })

    const options = wrapper.vm.options
    expect(options).toMatchObject({
      macros: {
        '\\blah': '\\frac{#}{#}',
      },
    })
  })

  it('merges global options', () => {
    const wrapper = mount(KatexElement, {
      props: {
        expression: '\\frac{a_i}{1+x}',
        displayMode: true,
        errorColor: '#fff',
        macros: {
          '\\blahblah': '\\frac{#}{#}',
        },
      },
      global: {
        plugins: [
          [
            VueKatex,
            {
              globalOptions: {
                displayMode: false,
                errorColor: '#000',
                macros: {
                  '\\blah': '\\frac{#}{#}',
                },
              },
            },
          ],
        ],
        shallow: true,
      },
    })

    const options = wrapper.vm.options
    expect(options).toMatchObject({
      macros: {
        '\\blah': '\\frac{#}{#}',
        '\\blahblah': '\\frac{#}{#}',
      },
      displayMode: true,
      errorColor: '#fff',
    })
  })

  it('props are mapped to options', () => {
    const displayMode = true
    const throwOnError = true
    const errorColor = '#ffffff'
    const macros = { '\\RR': '\\mathbb{R}' }
    const colorIsTextColor = true
    const maxSize = 100
    const maxExpand = 100
    const allowedProtocols = ['http', 'https']
    const strict = false

    const wrapper = mount(KatexElement, {
      props: {
        expression: '\\frac{a_i}{1+x}',
        displayMode,
        throwOnError,
        errorColor,
        macros,
        colorIsTextColor,
        maxSize,
        maxExpand,
        allowedProtocols,
        strict,
      },
      global: {
        plugins: [VueKatex],
        shallow: true,
      },
    })

    const options = wrapper.vm.options
    expect(options).toMatchObject({
      displayMode,
      throwOnError,
      errorColor,
      macros,
      colorIsTextColor,
      maxSize,
      maxExpand,
      allowedProtocols,
      strict,
    })
  })

  it('has correct root element - inline mode', () => {
    const wrapper = mount(KatexElement, {
      props: { expression: '\\frac{a_i}{1+x}' },
      global: {
        plugins: [VueKatex],
      },
    })
    expect(wrapper.find('span').exists()).toBe(true)
  })

  it('has correct root element - display mode', () => {
    const wrapper = mount(KatexElement, {
      props: {
        expression: '\\frac{a_i}{1+x}',
        displayMode: true,
      },
      global: {
        plugins: [VueKatex],
        shallow: true,
      },
    })
    expect(wrapper.find('div').exists()).toBe(true)
    expect(wrapper.html().substring(0, 4)).toBe('<div')
  })

  it('matches katex renderToString', () => {
    const expression = '\\frac{a_i}{1+x}'
    const wrapper = mount(KatexElement, {
      props: {
        expression,
      },
      global: {
        plugins: [VueKatex],
      },
    })
    const expectedInnerHtml = katex.renderToString(expression)
    expect(wrapper.html()).toContain(expectedInnerHtml)
  })
})
