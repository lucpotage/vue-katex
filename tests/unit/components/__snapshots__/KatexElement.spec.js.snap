// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KatexElement.vue > matches snapshot - display mode 1`] = `
VueWrapper {
  "_emitted": {},
  "_emittedByOrder": [],
  "isFunctionalComponent": undefined,
}
`;

exports[`KatexElement.vue > matches snapshot - inline mode 1`] = `
VueWrapper {
  "_emitted": {},
  "_emittedByOrder": [],
  "isFunctionalComponent": undefined,
}
`;

exports[`KatexElement.vue matches snapshot - display mode 1`] = `<div><span class="katex-display"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><msub><mi>a</mi><mi>i</mi></msub><mrow><mn>1</mn><mo>+</mo><mi>x</mi></mrow></mfrac></mrow><annotation encoding="application/x-tex">\\frac{a_i}{1+x}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.87689em;vertical-align:-0.7693300000000001em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:1.1075599999999999em;"><span style="top:-2.314em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord">1</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222222222222222em;"></span><span class="mord mathdefault">x</span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.677em;"><span class="pstrut" style="height:3em;"></span><span class="mord"><span class="mord"><span class="mord mathdefault">a</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.31166399999999994em;"><span style="top:-2.5500000000000003em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathdefault mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.7693300000000001em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span></div>`;

exports[`KatexElement.vue matches snapshot - inline mode 1`] = `<span><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mfrac><msub><mi>a</mi><mi>i</mi></msub><mrow><mn>1</mn><mo>+</mo><mi>x</mi></mrow></mfrac></mrow><annotation encoding="application/x-tex">\\frac{a_i}{1+x}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.114823em;vertical-align:-0.403331em;"></span><span class="mord"><span class="mopen nulldelimiter"></span><span class="mfrac"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.7114919999999999em;"><span style="top:-2.655em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight">1</span><span class="mbin mtight">+</span><span class="mord mathdefault mtight">x</span></span></span></span><span style="top:-3.23em;"><span class="pstrut" style="height:3em;"></span><span class="frac-line" style="border-bottom-width:0.04em;"></span></span><span style="top:-3.4101em;"><span class="pstrut" style="height:3em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight"><span class="mord mtight"><span class="mord mathdefault mtight">a</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3280857142857143em;"><span style="top:-2.357em;margin-left:0em;margin-right:0.07142857142857144em;"><span class="pstrut" style="height:2.5em;"></span><span class="sizing reset-size3 size1 mtight"><span class="mord mathdefault mtight">i</span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.143em;"><span></span></span></span></span></span></span></span></span></span></span><span class="vlist-s">​</span></span><span class="vlist-r"><span class="vlist" style="height:0.403331em;"><span></span></span></span></span></span><span class="mclose nulldelimiter"></span></span></span></span></span></span>`;
